File: 2265/Problem2265.cpp
#include <iostream>
#include <utility>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
 
class Solution {
public:
    int averageOfSubtree(TreeNode* root) {
        int solution = 0;
        calculateSumAndCount(root, solution);
        return solution;
    }

    pair<int, int> calculateSumAndCount(TreeNode* root, int &solution) {
        if(root == NULL)
            return make_pair(0, 0);

        pair<int, int> leftSubtree = calculateSumAndCount(root->left, solution);
        pair<int, int> rightSubtree = calculateSumAndCount(root->right, solution);

        int sum = root->val + leftSubtree.first + rightSubtree.first;
        int nodeCount = 1 + leftSubtree.second + rightSubtree.second;

        if(root->val == sum/nodeCount)
            solution++;

        return make_pair(sum, nodeCount);
    }
};

int main() {
    TreeNode* root = new TreeNode(4);
    root->left = new TreeNode(8);
    root->right = new TreeNode(5);
    root->left->left = new TreeNode(0);
    root->left->right = new TreeNode(1);
    root->left->right->left = nullptr;
    root->left->right->right = new TreeNode(6);

    Solution solution;
    int result = solution.averageOfSubtree(root);
    cout << "Number of nodes where the value is equal to the average of its subtree: " << result << std::endl;

    delete root->left->left;
    delete root->left->right->right;
    delete root->left->right;
    delete root->right;
    delete root->left;
    delete root;

    return 0;
}

File: 2265/Problem2265.py
class TreeNode:
    def __init__(self, val = 0, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right

class Solution: 
    def averageOfSubtree(self, root):
        solution = [0]
        self.calculateSumAndCount(root, solution)
        return solution[0]

    def calculateSumAndCount(self, root, solution):
        if root is None:
            return (0,0)
        
        leftSubtree = self.calculateSumAndCount(root.left, solution)
        rightSubtree = self.calculateSumAndCount(root.right, solution)

        sum = root.val + leftSubtree[0] + rightSubtree[0]
        nodeCount = 1 + leftSubtree[1] + rightSubtree[1]

        if root.val == sum/nodeCount:
            solution[0] += 1

        return (sum, nodeCount)

# Funzione per costruire l'albero specificato
def build_tree():
    # Costruzione dell'albero secondo la specifica [4,8,5,0,1,null,6]
    root = TreeNode(4)
    root.left = TreeNode(8)
    root.right = TreeNode(5)
    root.left.left = TreeNode(0)
    root.left.right = TreeNode(1)
    root.left.right.right = TreeNode(6)

    return root

# Funzione main
def main():
    # Costruzione dell'albero
    root = build_tree()

    # Creazione di un'istanza della classe Solution
    solution = Solution()

    # Calcolo del numero di nodi dove il valore è uguale alla media dei sottoalberi
    result = solution.averageOfSubtree(root)

    # Stampare il risultato
    print("Numero di nodi in cui il valore è uguale alla media dei sottoalberi:", result)

if __name__ == "__main__":
    main()  # Chiamata alla funzione main